### ISSUE TITLE
Keras Hub Model Conversion Fails

### ISSUE THREAD

**Issue by kossyrev-bg** [tensorflow/tensorflow#89904]
### 1. System information

- Ubuntu 18.04, python 3.12.3
- tf installed via `uv pip` 
- `tensorflow==2.18.1, keras==3.9.0`

### 2. Code

```
# %%
import os
os.environ['KERAS_BACKEND'] = 'tensorflow'  # Set the models from keras_hub to use tensorflow

# %%
import tensorflow as tf
import tensorflow.keras as keras
import keras_hub
from PIL import Image
from IPython.display import display
import numpy as np

# %%

# %%
# Input variables here
# See https://keras.io/keras_hub/presets/
MODEL_NAME = "efficientnet_b0_ra4_e3600_r224_imagenet"
# Path to training data
TRAINING_DATA_PATH = ...
# How many classes we want to classify
NUM_CLASSES = 2
## Training-related tuning
# How many GPUs to use, assumes they are named GPU:0, GPU:1, etc.
NUM_GPUS = 4
BATCH_SIZE = 32  # Per GPU
NUM_EPOCHS = 10
NUM_AUGS = 2  # Number of random augmentations to apply to each image
AUG_STRENGTH = 0.5  # How strong the augmentations should be
# Whether to quantize per-tensor or per-channel
PER_TENSOR_TFLITE = False 
IMAGE_SIZE = 224  
GLOBAL_BATCH = NUM_GPUS * BATCH_SIZE

# %%
# Create a MirroredStrategy.
gpus = [f"GPU:{i}" for i in range(NUM_GPUS)]
strategy = tf.distribute.MirroredStrategy(gpus)
print("Number of devices: {}".format(strategy.num_replicas_in_sync))

# Open a strategy scope.
with strategy.scope():
    # Everything that creates variables should be under the strategy scope.
    # In general this is only model construction & `compile()`.
    image_classifier = keras_hub.models.ImageClassifier.from_preset(
        MODEL_NAME,
        activation=None,  # outputs logits
        num_classes=NUM_CLASSES
    )
    image_classifier.compile(
        optimizer=keras.optimizers.Adam(
            learning_rate=1e-4, 
            clipnorm=1.0
        ),
        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),  # Consumes logits
        metrics=[keras.metrics.SparseCategoricalAccuracy()]
    )

# %%
# Create dataset
train_dataset, val_dataset = keras.preprocessing.image_dataset_from_directory(
    TRAINING_DATA_PATH,
    labels="inferred",
    label_mode="int",
    batch_size=None,
    image_size=(IMAGE_SIZE, IMAGE_SIZE),
    shuffle=True,
    validation_split=0.2,
    subset="both",
    interpolation="bilinear",
    verbose=True,
    seed=42
)

# Define simple augmenter
aug_layer = keras.layers.RandAugment(value_range=(0, 255), num_ops=NUM_AUGS, factor=AUG_STRENGTH)
map_fn = lambda x, y: (aug_layer(x, training=True), y)

# Prefetch the data, adding augmentation to the training dataset
train_dataset = train_dataset.batch(GLOBAL_BATCH, drop_remainder=True).map(map_fn, num_parallel_calls=tf.data.AUTOTUNE).prefetch(tf.data.AUTOTUNE)
val_dataset = val_dataset.batch(GLOBAL_BATCH, drop_remainder=True).prefetch(tf.data.AUTOTUNE)

# %%

# %%
# Freeze the base model to only retrain the head (optional)
# image_classifier.backbone.trainable = False
# Train the model on all available devices.
image_classifier.fit(train_dataset, epochs=NUM_EPOCHS, validation_data=val_dataset)

# Test the model on all available devices.
# TODO: Need a test dataset
# image_classifier.evaluate(test_dataset)
image_classifier.save(f"{MODEL_NAME}.keras")

# %%
# Test the trained model on 10 validation images
# This works fine, getting at least 8 correct
for imgs, label in val_dataset.take(1):
    for i in range(10):
        img = imgs[i:i+1]
        preds = image_classifier.predict(img)
        print(f"Raw preds: {preds}, Predicted: {preds.argmax()}, Actual: {label[i]}")

# %%
# Remove preprocessor
model_preprocessor = image_classifier.preprocessor
image_classifier.preprocessor = None

# %%
### Create end-to-end model that takes in UINT8 image of size 224 ###
# Modify preprocessor, because GPU doesn't like bicubic? idk
cfg = model_preprocessor.get_config()
cfg["image_converter"]["config"]["interpolation"] = "bilinear"
new_preproc = keras_hub.models.ImageClassifierPreprocessor.from_config(cfg)
# Input layer is (1, 224, 224, 3) uint8
input_layer = keras.layers.Input(shape=(224, 224, 3), dtype=tf.uint8, batch_size=1)
# Convert to float 32 like the preprocessor expects (I'd like to skip this if possible,
# because it might be part of why the tflite conversion is failing,
# but I don't know how without dismantling the preprocessor)
float_input = keras.layers.Lambda(lambda x: tf.cast(x, tf.float32))(input_layer)
# Preprocess the image
preprocessed_input = new_preproc(float_input, training=False)
# Run through the model - no double-preprocessing since we removed it
logits = image_classifier(preprocessed_input, training=False)
# Combine into one model
full_model = keras.models.Model(inputs=float_input, outputs=logits)
# Compile the model (unclear if necessary)
full_model.compile(
    optimizer=keras.optimizers.Adam(1e-4),
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),  # Consumes logits
    metrics=[keras.metrics.SparseCategoricalAccuracy()]
)
full_model.trainable = False

# %%
# Test the new model on 10 validation images. This works too, roughly same
# accuracy as above
num_correct = 0
num_wrong = 0
for imgs, label in val_dataset.take(1):
    for i in range(GLOBAL_BATCH):
        img = imgs[i:i+1]
        # preds = full_model.predict(tf.cast(img, tf.uint8))
        preds = full_model.predict(img)
        print(f"Raw preds: {preds}, Predicted: {preds.argmax()}, Actual: {label[i]}")
        if preds.argmax() == label[i]:
            num_correct += 1
        else:
            num_wrong += 1
print(f"Correct: {num_correct}, Wrong: {num_wrong}")

# %%
# Use data from the validation dataset to calibrate the quantized model
def representative_data_gen():
    batch_count = (300 // GLOBAL_BATCH) + 1
    run = 0
    for x, _ in val_dataset.take(batch_count):  # Use a few samples
        run += 1
        print(f"Running batch {run}/{batch_count}")
        # Pull one image at a time
        for i in range(GLOBAL_BATCH):
            img = x[i:i+1]
            yield [tf.cast(img, tf.uint8)]
            yield [img]
# Convert to tflite, using per-tensor quantization for ARTPEC-8
converter = tf.lite.TFLiteConverter.from_keras_model(full_model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.representative_dataset = representative_data_gen
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.float32
if PER_TENSOR_TFLITE:
    converter._experimental_disable_per_channel = True
tflite_quant_model = converter.convert()
with open(f"{MODEL_NAME}.tflite", "wb") as f:
    f.write(tflite_quant_model)

# %%
# Run the TFLite model on some test data
interpreter = tf.lite.Interpreter(model_content=tflite_quant_model)
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
input_index = input_details[0]["index"]
output_index = output_details[0]["index"]
interpreter.allocate_tensors()

# %%
num_correct = 0
num_wrong = 0
for x, label in val_dataset.take(3):
    for i in range(GLOBAL_BATCH):
        img = x[i:i+1]
        # Need to cast to uint8 because tflite model expects it
        interpreter.set_tensor(input_index, tf.cast(img, tf.uint8))
        interpreter.invoke()
        output_data = interpreter.get_tensor(output_index)
        print(f"Raw preds: {output_data}, Predicted: {output_data.argmax()}, Actual: {label[i]}")
        if output_data.argmax() == label[i]:
            num_correct += 1
        else:
            num_wrong += 1
print(f"{num_correct=}, {num_wrong=}")
```

### 3. Failure after conversion
Model accuracy drastically drops after conversion:
```
Raw preds: [[-0.08086799  0.20216998]], Predicted: 1, Actual: 0
Raw preds: [[ 0.       -0.040434]], Predicted: 0, Actual: 1
Raw preds: [[-1.0512838   0.48520795]], Predicted: 1, Actual: 0
Raw preds: [[-1.0108498   0.36390597]], Predicted: 1, Actual: 1
Raw preds: [[-0.9704159   0.12130199]], Predicted: 1, Actual: 0
Raw preds: [[-0.72781193  0.24260397]], Predicted: 1, Actual: 1
Raw preds: [[-0.08086799 -0.24260397]], Predicted: 0, Actual: 0
Raw preds: [[ 0.16173598 -0.040434  ]], Predicted: 0, Actual: 0
Raw preds: [[-1.1321518  0.040434 ]], Predicted: 1, Actual: 0
...
Raw preds: [[-0.24260397  0.12130199]], Predicted: 1, Actual: 1
Raw preds: [[-0.76824594 -0.08086799]], Predicted: 1, Actual: 0
Raw preds: [[-0.36390597  0.24260397]], Predicted: 1, Actual: 1

num_correct=196, num_wrong=188
```

**Comment by mihaimaruseac** [comment]
This belongs in keras repos.

**Comment by kossyrev-bg** [comment]
I found the problem with the model setup. However, after the tflite conversion, the model accuracy is still drastically lower than it was before. I can update the code snippet to show the new setup, but it's basically just combining preprocessing with the model into one all-in-one model. I would still like help with why tflite conversion is reducing accuracy by so much.

**Comment by mihaimaruseac** [comment]
Oh, I misunderstood (from the title), this is not about Keras but model conversion to TFLite. Can you minimize this to a model as small as possible?

**Comment by kossyrev-bg** [comment]
This is an efficientnet model from the keras hub: https://www.kaggle.com/models/keras/efficientnet/keras/efficientnet_b0_ra4_e3600_r224_imagenet/

It's about 5MB after quantization to tflite. Is that too big? Do you need the keras model and the tflite model, and do you need the data?

**Comment by gaikwadrahul8** [comment]
Hi, @kossyrev-bg 
I apologize for the delay in response, I am able to replicate the same behavior from my end with sample flowers dataset with your provided code snippet here is [gist-file](https://colab.sandbox.google.com/gist/gaikwadrahul8/ed224e71bc246b43f0bfe9d680e7d078/tflite-89904.ipynb) for reference so we will have to dig more into this issue and will update you, thank you for bringing this issue to our attention. Thank you for your cooperation.

```
Original model predictions:
1/1 ━━━━━━━━━━━━━━━━━━━━ 3s 3s/step
Raw preds: [[-2.7760098  3.1728508]], Predicted: 1, Actual: 1
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 326ms/step
Raw preds: [[-0.55148685  0.8257771 ]], Predicted: 1, Actual: 1
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 280ms/step
Raw preds: [[ 3.8654664 -4.097068 ]], Predicted: 0, Actual: 0
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 299ms/step
Raw preds: [[ 15.905437 -15.347878]], Predicted: 0, Actual: 0
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 313ms/step
Raw preds: [[-0.43868783  1.6053506 ]], Predicted: 1, Actual: 1
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 247ms/step
Raw preds: [[ 13.696747 -12.681302]], Predicted: 0, Actual: 0
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 199ms/step
Raw preds: [[-1.2708153   0.60328835]], Predicted: 1, Actual: 1
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 188ms/step
Raw preds: [[0.7198984  0.65562236]], Predicted: 0, Actual: 1
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 207ms/step
Raw preds: [[-1.3494838  2.3646579]], Predicted: 1, Actual: 1
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 192ms/step
Raw preds: [[-1.334792   1.3840239]], Predicted: 1, Actual: 1
```

```
TFLite model predictions:
TFLite output: [[-2.4290873e+13  2.2772695e+13]], Predicted class: 1
TFLite output: [[-6.0727183e+12  3.0363592e+12]], Predicted class: 1
TFLite output: [[0. 0.]], Predicted class: 0
TFLite output: [[-2.1254514e+13  1.9736335e+13]], Predicted class: 1
TFLite output: [[-4.2509028e+13  2.1254514e+13]], Predicted class: 1
TFLite output: [[-2.2772695e+13  2.1254514e+13]], Predicted class: 1
TFLite output: [[-7.5908980e+12  1.2145437e+13]], Predicted class: 1
TFLite output: [[-3.1881771e+13  2.1254514e+13]], Predicted class: 1
TFLite output: [[-2.4290873e+13  2.1254514e+13]], Predicted class: 1
TFLite output: [[0. 0.]], Predicted class: 0
TFLite output: [[2.4290873e+13 3.4918130e+13]], Predicted class: 1
TFLite output: [[-1.3663616e+13  1.5181796e+12]], Predicted class: 1
TFLite output: [[-1.5181796e+13  2.2772695e+13]], Predicted class: 1
TFLite output: [[0. 0.]], Predicted class: 0
TFLite output: [[-3.4918130e+13  2.7327233e+13]], Predicted class: 1
TFLite output: [[-1.0627257e+13  2.1254514e+13]], Predicted class: 1
```

**Comment by kossyrev-bg** [comment]
Okay @gaikwadrahul8 the code has been updated, with new example logits, thank you for your help!

**Comment by kossyrev-bg** [comment]
I'll go ahead and update the gist file as well, if I can

**Comment by gaikwadrahul8** [comment]
Hi @kossyrev-bg
To facilitate further investigation and debugging of the current issue, we would greatly appreciate it if you could consider updating the existing gist file or creating a new Google Colab notebook including a sample flowers dataset within this resource would be particularly helpful for our team to reproduce the reported behavior and analyze the underlying cause effectively.

Thank you for your cooperation.

**Comment by gaikwadrahul8** [comment]
Hi, @kossyrev-bg 
I see you're using `int8`, TensorFlow Lite offers other post-training quantization options like `float16` quantization. This offers a smaller model size reduction and speedup compared to `int8` but often preserves accuracy much better so could you please give it try with `float16` quantization something like below and see are you getting better predictions as compared to `int8` quantization ? 

```
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.target_spec.supported_types = [tf.float16] 
    tflite_model = converter.convert()
    with open(save_path_tflite, "wb") as f:
        f.write(tflite_model)
```

Thank you for your cooperation.

**Comment by mihaimaruseac** [comment]
Oh, quantization is expected to lose precision, you have fewer bits to represent the numbers from the weights and inputs/outputs.

**Comment by kossyrev-bg** [comment]
I am aware, but I didn't think it would be quite this bad. Would a larger model perhaps help with this penalty?

**Comment by gaikwadrahul8** [comment]
Hi, @kossyrev-bg
Larger models with more parameters can sometimes mitigate the effects of quantization because they have more capacity to learn and generalize. However, this is not a guaranteed solution and can lead to increased computational requirements.

`int8` quantization is necessary for your edge system it often requires careful tuning and adjustments to maintain accuracy. Implementing quantization aware training is likely your best bet for improving performance please refer this [Quantization aware training in Keras example](https://www.tensorflow.org/model_optimization/guide/quantization/training_example). Additionally fine-tuning the model after quantization and ensuring a robust representative dataset can also help mitigate accuracy loss.

Thank you for your cooperation.

**Comment by kossyrev-bg** [comment]
I'll try out the example and update the gist accordingly, thank you

**Comment by kossyrev-bg** [comment]
It seems that `keras_hub` does not play nicely with the `keras` from `from tensorflow_model_optimization.python.core.keras.compat import keras`

If there's no easy way around this, I guess I'll have to abandon using kerashub for my models :/

**Comment by kossyrev-bg** [comment]
Having better luck moving over to tensorflow_hub. Closing this as no longer relevant.
